File: src/main/scala/io/izzel/taboolib/cronus/bridge/database/BridgeDatabase.java
package io.izzel.taboolib.cronus.bridge.database;

import com.google.common.collect.Maps;
import com.mongodb.ConnectionString;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoDatabase;

import java.util.Map;

public class BridgeDatabase {

   private final String client;
   private final String database;
   private final MongoClient mongoClient;
   private final MongoDatabase mongoDatabase;
   private final Map<String, BridgeCollection> collectionMap = Maps.newConcurrentMap();

   public BridgeDatabase(String client, String database) {
      this.client = client;
      this.database = database;
      this.mongoClient = MongoClients.create(new ConnectionString(this.client));
      this.mongoDatabase = this.mongoClient.getDatabase(this.database);
   }

   public BridgeCollection get(String collection) {
      return this.collectionMap.computeIfAbsent(collection, i -> new BridgeCollection(this, collection));
   }

   public BridgeCollection get(String collection, IndexType indexType) {
      return this.collectionMap.computeIfAbsent(collection, i -> new BridgeCollection(this, collection, indexType));
   }

   public void release(String collection) {
      this.collectionMap.remove(collection);
   }

   public void releaseId(String collection, String id) {
      BridgeCollection bridgeCollection = this.collectionMap.get(collection);
      if (bridgeCollection != null) {
         bridgeCollection.getDataMap().remove(id);
      }
   }

   public void close() {
      this.mongoClient.close();
   }

   public String getClient() {
      return this.client;
   }

   public String getDatabase() {
      return this.database;
   }

   public MongoClient getMongoClient() {
      return this.mongoClient;
   }

   public MongoDatabase getMongoDatabase() {
      return this.mongoDatabase;
   }

   public Map<String, BridgeCollection> getCollectionMap() {
      return collectionMap;
   }
}

File: module-database-mongodb/src/main/kotlin/taboolib/module/database/bridge/BridgeDatabase.kt
package taboolib.module.database.bridge

import com.mongodb.ConnectionString
import com.mongodb.client.MongoClient
import com.mongodb.client.MongoClients
import com.mongodb.client.MongoDatabase
import java.util.concurrent.ConcurrentHashMap

class BridgeDatabase(val client: String, val database: String) {

    val mongoClient: MongoClient = MongoClients.create(ConnectionString(client))
    val mongoDatabase: MongoDatabase = mongoClient.getDatabase(database)
    val collectionMap = ConcurrentHashMap<String, BridgeCollection>()

    fun getCollection(collection: String, indexType: Index = Index.NONE): BridgeCollection {
        return collectionMap.computeIfAbsent(collection) { BridgeCollection(this, collection, indexType) }
    }

    fun release(collection: String) {
        collectionMap.remove(collection)
    }

    fun releaseId(collection: String, id: String?) {
        val bridgeCollection = collectionMap[collection]
        bridgeCollection?.dataMap?.remove(id)
    }

    fun close() {
        mongoClient.close()
    }
}